### PR TITLE
Adding policies

### PR DESCRIPTION
**Pull request by sriramss**
[Kyverno logs.docx](https://github.com/Cray-HPE/cray-kyverno-policies/files/8504997/Kyverno.logs.docx)
## Summary and Scope

Adding kyverno mutation and validation policies.

## Issues and Related PRs

Resolves [CASMPET-5448](https://jira-pro.its.hpecorp.net:8443/browse/CASMPET-5448)

### Tested on:

  * Starlord
  * Mug
  * Virtual Shasta

### Test description:

Respective workloads were restarted after enforcing the policies and they have come up successfully.

Eg:
1. Fetch the kyverno policy

ncn-m001:~ # kubectl get pol -A
```
NAMESPACE   NAME                                                BACKGROUND   ACTION   READY
services    pod-sec-ctxt-capmc                                  true         audit    true
services    validate-pod-sec-ctxt-capmc                         true         audit    true

```
2. Resource to which policy is applied

ncn-m001:~ # kubectl get pods -n services | grep capmc
```
cray-capmc-86c5d7795f-fqjh4                                    2/2     Running             0         19d
cray-capmc-59965cb595-f252t                                    2/2     Running            56         54d
cray-capmc-59965cb595-znk8p                                    2/2     Running            56         54d
```

3. Fetch the manifest before policy enforcement

    image: artifactory.algol60.net/csm-docker/stable/cray-capmc:1.32.0
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /app-health/cray-capmc/livez
        port: 15020
        scheme: HTTP
      initialDelaySeconds: 10
      periodSeconds: 20
      successThreshold: 1
      timeoutSeconds: 1
    name: cray-capmc
    ports:
    - containerPort: 27777
      name: http
      protocol: TCP
    ...
    **securityContext:
      runAsGroup: 65534
      runAsNonRoot: true
      runAsUser: 65534**

4. Fetch the validation report before policy enforcement

ncn-m001:~ # kubectl get polr -A -o wide
```
NAMESPACE      NAME                   KIND   NAME   PASS   FAIL   WARN   ERROR   SKIP   AGE
istio-system   polr-ns-istio-system                 0      0      0      0       0      52d
opa            polr-ns-opa                          0      0      0      0       0      52d
operators      polr-ns-operators                    0      0      0      0       0      53d
services       polr-ns-services                     0      3      0      0       0      53d
spire          polr-ns-spire                        0      0      0      0       0      32d
vault          polr-ns-vault                        0      0      0      0       0      49d
```

5. Restart the pod by deleting it to apply the policy
           
ncn-m001:~ # kubectl delete pod -n services cray-capmc-86c5d7795f-fqjh4
pod "cray-capmc-86c5d7795f-fqjh4" deleted
 
6. Check the pod bring up after policy is enforced
 
ncn-m001:~ # kubectl get pods -n services | grep capmc
```
cray-capmc-86c5d7795f-bsfj5                                    2/2     Running            0          2m50s
cray-capmc-59965cb595-f252t                                    2/2     Running            56         54d
cray-capmc-59965cb595-znk8p                                    2/2     Running            56         54d

```
7. Manifest after policy enforcement

    image: artifactory.algol60.net/csm-docker/stable/cray-capmc:1.32.0
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /app-health/cray-capmc/livez
        port: 15020
        scheme: HTTP
      initialDelaySeconds: 10
      periodSeconds: 20
      successThreshold: 1
      timeoutSeconds: 1
    name: cray-capmc
    ports:
    - containerPort: 27777
      name: http
      protocol: TCP
    ...
    **securityContext:
      allowPrivilegeEscalation: false
      privileged: false
      runAsGroup: 65534
      runAsNonRoot: true
      runAsUser: 65534**

8. Fetch the validation report after policy enforcement

ncn-m001:~ # kubectl get polr -A -o wide
```
NAMESPACE      NAME                   KIND   NAME   PASS   FAIL   WARN   ERROR   SKIP   AGE
istio-system   polr-ns-istio-system                 0      0      0      0       0      52d
opa            polr-ns-opa                          0      0      0      0       0      52d
operators      polr-ns-operators                    0      0      0      0       0      53d
services       polr-ns-services                     1      2      0      0       0      53d
spire          polr-ns-spire                        0      0      0      0       0      32d
vault          polr-ns-vault                        0      0      0      0       0      49d
```
Please refer to the attached file for more test logs.

## Risks and Mitigations

NA

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: charts/kyverno-policies/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+appVersion: v1.6.0
+icon: https://github.com/kyverno/kyverno/raw/main/img/logo.png
+description: Kubernetes Pod Security Standards implemented as Kyverno policies
+keywords:
+  - kubernetes
+  - nirmata
+  - policy agent
+  - validating webhook
+  - admissions controller
+home: https://kyverno.io/policies/
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/Chart.yaml:13]: This is not related to the upstream Kyverno policies repo. We shouldn't reference it here.

File: charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  namespace: services
+  annotations:
+    policies.kyverno.io/title: Mutate pod and container security contexts
+    policies.kyverno.io/category: Pod Security Standards (Baseline)
+    {{- if .Values.podSecuritySeverity }}
+    policies.kyverno.io/severity: {{ .Values.podSecuritySeverity }}
+    {{- end }}
+    kyverno.io/kyverno-version: 1.6.0
+    policies.kyverno.io/minversion: 1.6.0
+    kyverno.io/kubernetes-version: "1.22-1.23"
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml:17]: CSM 1.2.0 will be running K8S 1.20. Not sure if there is a K8S update planned for 1.5 yet. 

So, why is this range here?

File: charts/kyverno-policies/values.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Supported- baseline/restricted/privileged/custom
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/values.yaml:1]: This looks like boilerplate from the upstream kyverno policies repository. As we are not using the upstream, what is the value of this here?

File: charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire.yaml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    policies.kyverno.io/category: Pod Security Standards (Baseline)
+    {{- if .Values.podSecuritySeverity }}
+    policies.kyverno.io/severity: {{ .Values.podSecuritySeverity }}
+    {{- end }}
+    kyverno.io/kyverno-version: 1.6.0
+    policies.kyverno.io/minversion: 1.6.0
+    kyverno.io/kubernetes-version: "1.22-1.23"
+    policies.kyverno.io/subject: Pod
+    policies.kyverno.io/description: >-
+      Example, highly restrictive mutate policy setting pod and container security context
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire.yaml:20]: These aren't really examples once they are checked, each policy has a specific purpose.

**Review by jeremy-duckworth — CHANGES_REQUESTED** [review]

File: charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    policies.kyverno.io/subject: Pod
+    policies.kyverno.io/description: >-
+      Example, highly restrictive mutate policy setting pod and container security context
```
Comment (spillerc-hpe) [inline charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml:20]: The description should reflect what the policy does or at least remove the word `Example` from the text.

File: charts/kyverno-policies/Chart.yaml (hunk 1 above)
Comment (sriramss) [reply charts/kyverno-policies/Chart.yaml:13]: Okay

File: charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml (hunk 2 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml:17]: I had copied it from the upstream repo. Let me remove it!

File: charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml (hunk 5 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml:20]: okay

File: charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire.yaml (hunk 4 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire.yaml:20]: okay

File: charts/kyverno-policies/values.yaml (hunk 3 above)
Comment (sriramss) [reply charts/kyverno-policies/values.yaml:1]: We don't have any categories as such. Can we just keep baseline or remove it as well?

File: charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    policies.kyverno.io/title: Mutate pod and container security contexts
+    policies.kyverno.io/category: Pod Security Standards (Baseline)
+    {{- if .Values.podSecuritySeverity }}
+    policies.kyverno.io/severity: {{ .Values.podSecuritySeverity }}
+    {{- end }}
+    kyverno.io/kyverno-version: 1.6.0
+    policies.kyverno.io/minversion: 1.6.0
+    policies.kyverno.io/subject: Pod
+    policies.kyverno.io/description: >-
+      Highly restrictive mutate policy setting container security context
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml:19]: Please set specific descriptions and other annotations that you think will help a developer or operator.

File: charts/kyverno-policies/templates/_helpers.tpl [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+app.kubernetes.io/component: kyverno
+app.kubernetes.io/instance: {{ .Release.Name }}
+app.kubernetes.io/managed-by: {{ .Release.Service }}
+app.kubernetes.io/name: {{ template "kyverno-policy.name" . }}
+app.kubernetes.io/part-of: {{ template "kyverno-policy.name" . }}
+app.kubernetes.io/version: "{{ .Chart.Version }}"
+helm.sh/chart: {{ include "kyverno-policy.chart" . }}
+{{- end -}}
+
+{{/* Set if a baseline policy is managed */}}
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/templates/_helpers.tpl:26]: What is the value/function of this?

File: charts/kyverno-policies/templates/pod-sec-ctxt-customer-access-ingress.yaml [hunk 8]
```diff
@@ -0,0 +1,42 @@
+{{- $name := "pod-sec-ctxt-customer-access-ingress" }}
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/templates/pod-sec-ctxt-customer-access-ingress.yaml:1]: If this is an istio ingress gateway, remove. Same for similar workloads. Remove the validation policies as well.

File: charts/kyverno-policies/values.yaml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+podSecurityStandard: baseline
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/values.yaml:1]: Need some context on why this and the template boilerplate exists.

File: charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    mutate:
+      foreach:
+      - list: "request.object.spec.containers"
+        patchStrategicMerge:
+          spec:
+            containers:
+            - (name): |-
+                {{`{{element.name}}`}}
+              securityContext:
+                +(allowPrivilegeEscalation): false
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml:40]: The +() only sets the tag if it is not present, https://kyverno.io/docs/writing-policies/mutate/#conditional-logic-using-anchors. This is not the behavior we want. I believe you'll want to remove the conditional logic. 

For non-exceptions, a workload is allowed to override runAsGroup and runAsUser. No override will be allowed for runAsNonRoot (true), allowPrivilegeEscalation (false), nor privileged (false), unless by documented exception (and there shouldn't be any in this PR).

**Review by jeremy-duckworth — CHANGES_REQUESTED** [review]

File: charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml (hunk 10 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml:40]: okay

File: charts/kyverno-policies/values.yaml (hunk 9 above)
Comment (sriramss) [reply charts/kyverno-policies/values.yaml:1]: The upstream repo has different categories -> https://github.com/kyverno/policies/tree/main/pod-security  Let's decide if we need the same

File: charts/kyverno-policies/templates/pod-sec-ctxt-customer-access-ingress.yaml (hunk 8 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-customer-access-ingress.yaml:1]: We have already filtered out the istio-gateway related resources. This has nothing to do with it!

File: charts/kyverno-policies/templates/_helpers.tpl (hunk 7 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/_helpers.tpl:26]: We are defining a condition here to enable the policy based on the policy type. We have just defined one(baseline) policy type in values.yaml . Hence, this condition is redundant as of now as all the policies in the folder get enforced.

File: charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml (hunk 6 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-capmc.yaml:19]: okay

File: charts/kyverno-policies/templates/pod-sec-ctxt-customer-access-ingress.yaml (hunk 8 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-customer-access-ingress.yaml:1]: okay.

File: charts/kyverno-policies/values.yaml (hunk 9 above)
Comment (sriramss) [reply charts/kyverno-policies/values.yaml:1]: okay

File: CHANGELOG.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+# Changelog
+
+## [v1.0.0] - 2022-04-18
+
+Inclusion of kyverno Policies in CSM 1.2.5 Patch
```
Comment (jeremy-duckworth) [inline CHANGELOG.md:5]: Change this to not include the CSM release. Just note that it is the initial version, what is actually in it.

File: charts/kyverno-policies/templates/NOTES.txt [hunk 12]
```diff
@@ -0,0 +1,5 @@
+Thank you for installing {{ .Chart.Name }} {{ .Chart.Version }} 😀
+
+We have installed the "{{ .Values.podSecurityStandard }}" profile of Pod Security Standards.
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/templates/NOTES.txt:3]: Strike reference of pod security standards.

File: .github/workflows/charts-lint-test-scan.yml [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+name: Lint, test, and scan Helm charts
+on:
+  pull_request:
+    branches:
+      - master
```
Comment (jeremy-duckworth) [inline .github/workflows/charts-lint-test-scan.yml:28]: main?

**Review by jeremy-duckworth — CHANGES_REQUESTED** [review]
Very close, need to fix branch ref in GH action workflow changelog nit.

File: charts/kyverno-policies/templates/NOTES.txt (hunk 12 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/NOTES.txt:3]: okay

File: CHANGELOG.md (hunk 11 above)
Comment (sriramss) [reply CHANGELOG.md:5]: okay

**Comment by sriramss** [comment]
> Very close, need to fix branch ref in GH action workflow changelog nit.

Done!

File: charts/kyverno-policies/Chart.yaml [hunk 14]
```diff
@@ -8,5 +8,8 @@ keywords:
   - policy agent
   - validating webhook
   - admissions controller
+maintainers:
+  - name: sriramss
+    url: https://kyverno.io/
```
[the remark this answers is not in the document]
Comment (sriramss) [reply charts/kyverno-policies/Chart.yaml:13]: okay

**Comment by sriramss** [comment]
> Remove upstream URL in chart maintainers section.

Done

File: charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire-server.yaml [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+      message: >-
+        Running as root is not allowed, Privilege escalation is disallowed and Privileged mode is disallowed. The fields
+        spec.containers[*].securityContext.runAsNonRoot must be set to `true`, spec.containers[*].securityContext.allowPrivilegeEscalation
+        must be set to `false` and spec.containers[*].securityContext.privileged must be unset or set to `false`.
+      pattern:
+        spec:
+          containers:
+          - (name): "*"
+            securityContext:
+              runAsNonRoot: true
```
Comment (jeremy-duckworth) [inline charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire-server.yaml:39]: Notably for the runasuser stuff, we're mutating a set of policies, but not verifying the same policy. Thinking we may not need the runasuser/group mutations?

**Review by jeremy-duckworth — CHANGES_REQUESTED** [review]

File: charts/kyverno-policies/templates/pod-sec-ctxt-ims.yaml [hunk 16]
```diff
@@ -0,0 +1,39 @@
+{{- $name := "pod-sec-ctxt-ims" }}
```
[the remark this answers is not in the document]
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-ims.yaml:1]: No, it doesn't have any exception!

File: charts/kyverno-policies/templates/pod-sec-ctxt-customer-management-ingress.yaml [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+    mutate:
+      foreach:
+      - list: "request.object.spec.containers"
+        patchStrategicMerge:
+          spec:
+            containers:
+            - (name): |-
+                {{`{{element.name}}`}}
+              securityContext:
+                +(runAsUser): 2000
```
[the remark this answers is not in the document]
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-customer-management-ingress.yaml:37]: When we looked into the shell, the container was already running with UID 2000. So, we used the same!

File: charts/kyverno-policies/templates/pod-sec-ctxt-spire.yaml [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    mutate:
+      foreach:
+      - list: "request.object.spec.containers"
+        patchStrategicMerge:
+          spec:
+            containers:
+            - (name): |-
+                {{`{{element.name}}`}}
+              securityContext:
+                +(runAsUser): 65534
```
[the remark this answers is not in the document]
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-spire.yaml:37]: We need to specify non-root  UID right?

File: charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire-server.yaml (hunk 15 above)
Comment (sriramss) [reply charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire-server.yaml:39]: No, we can add that too! Just to validate things, I thought runsAsNonRoot is enough. Should I add runasuser/group also?

File: charts/kyverno-policies/templates/pod-sec-ctxt-spire.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (jeremy-duckworth) [reply charts/kyverno-policies/templates/pod-sec-ctxt-spire.yaml:37]: Is the workload already running with this UID? This is an optional set, if the workload doesn't already provide the setting. 

Since we're forcing runAsNonRoot, the workload will be forced to run with a non-UID 0 user. See https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core.

File: charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire-server.yaml (hunk 15 above)
Comment (jeremy-duckworth) [reply charts/kyverno-policies/templates/validation/validate-pod-sec-ctxt-spire-server.yaml:39]: See recent comments. I think we should probably drop the runAsUser mutations.

File: charts/kyverno-policies/templates/pod-sec-ctxt-ims.yaml (hunk 16 above)
[the remark this answers is not in the document]
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-ims.yaml:1]: Yes, let's keep it for now as the functionality test has passed.

File: charts/kyverno-policies/templates/pod-sec-ctxt-customer-management-ingress.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-customer-management-ingress.yaml:37]: Shall I remove it in all the places where we have specified UIDs?

File: charts/kyverno-policies/templates/pod-sec-ctxt-spire.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (sriramss) [reply charts/kyverno-policies/templates/pod-sec-ctxt-spire.yaml:37]: So, shall we remove the UID setting in all the policies?

**Review by jeremy-duckworth — APPROVED** [review]

**Review by yarlagadda-srinivasa-rao — APPROVED** [review]
If time permits, as we have included minor changes, its good to capture latest sanity log in the PR from 1.2.6 installed system.
